### PR TITLE
[mlir][complex] Prevent underflow in complex.abs (#79786)

### PR DESCRIPTION
**Pull request by Lewuathe**
Fix the issue found in the previous change. https://github.com/llvm/llvm-project/commit/70fb96a28678c488326a135fe298175b9a3a7657 It is necessary to use the absolute value in the case of either real or imag part is zero as shown in [the original algorithm](https://dl.acm.org/doi/pdf/10.1145/363717.363780). 

See: https://github.com/llvm/llvm-project/pull/79786

### CODE REVIEW

File: mlir/lib/Conversion/ComplexToStandard/ComplexToStandard.cpp [hunk 1]
```diff
@@ -26,29 +26,59 @@ namespace mlir {
...
+    Value realDivImag = b.create<arith::DivFOp>(real, imag, fmf.getValue());
+    Value realSq =
+        b.create<arith::MulFOp>(realDivImag, realDivImag, fmf.getValue());
+    Value realSqPlusOne = b.create<arith::AddFOp>(realSq, one, fmf.getValue());
+    Value realSqrt = b.create<math::SqrtOp>(realSqPlusOne, fmf.getValue());
+    Value imagAbs = b.create<math::AbsFOp>(imag, fmf.getValue());
+    Value absReal = b.create<arith::MulFOp>(realSqrt, imagAbs, fmf.getValue());
+
+    rewriter.replaceOpWithNewOp<arith::SelectOp>(
+        op, realIsZero, imagAbs,
```
Comment (Lewuathe) [inline mlir/lib/Conversion/ComplexToStandard/ComplexToStandard.cpp:75]: Use `imagAbs` if real is zero.

File: mlir/lib/Conversion/ComplexToStandard/ComplexToStandard.cpp [hunk 2]
```diff
@@ -26,29 +26,59 @@ namespace mlir {
...
+        b.create<arith::SelectOp>(
+            imagIsZero, realAbs,
```
Comment (Lewuathe) [inline mlir/lib/Conversion/ComplexToStandard/ComplexToStandard.cpp:77]: Use `realAbs` if imag is zero.

File: mlir/test/Integration/Dialect/Complex/CPU/correctness.mlir [hunk 3]
```diff
@@ -300,5 +321,36 @@ func.func @entry() {
...
+    (0.0, 6.0),
+    // CHECK-NEXT:  6
+    (7.0, 8.0),
+    // CHECK-NEXT:  10.6301
+    (-1.0, -1.0),
+    // CHECK-NEXT: 1.414
+    (-1.0e300, -1.0e300),
+    // CHECK-NEXT:  1.41421e+300
+    (-1.0, 0.0),
+    // CHECK-NEXT:  1
```
Comment (Lewuathe) [inline mlir/test/Integration/Dialect/Complex/CPU/correctness.mlir:343]: Add test case the imag part is zero.

File: mlir/test/Integration/Dialect/Complex/CPU/correctness.mlir [hunk 4]
```diff
@@ -300,5 +321,36 @@ func.func @entry() {
...
+    (0.0, -1.0)
+    // CHECK-NEXT:  1
```
Comment (Lewuathe) [inline mlir/test/Integration/Dialect/Complex/CPU/correctness.mlir:345]: Add test case where the real is zero.

**Review by Lewuathe — COMMENTED** [review]
@d0k @matthias-springer @joker-eph Sorry for asking you to review the same change couple of times. I have detected and fixed the cause of the previous negative value case error. Could you review this when you get a chance?

**Review by d0k — APPROVED** [review]

**Comment by Lewuathe** [comment]
@d0k Thank you so much for reviewing.

**Pull request merged**